Added joystick settings

breiler opened the pull request:
Added settings in the joystick plugin making it possible to change mappings of the controller buttons and axises. The actions that can be mapped comes from Netbeans. Added a live preview of the pressed buttons.

![Screen Recording 2020-04-07 at 16 46 00 2020-04-07 16_51_07](https://user-images.githubusercontent.com/8962024/78812659-0f960280-79cc-11ea-82ef-1e2ba5f464c4.gif)

jiffyjeff commented:
Thanks for the work on this!  I built this branch and tested using a Logitech Rumblepad 2.  Two issues stand out:

1. The Y and Z axis directions were reversed for the analog joysticks.  (Digital dpads were fine.)  I see an intentional flip of Y values being done in jamepad's ControllerIndex.getAxisState().  There's a related github issue: https://github.com/williamahartman/Jamepad/issues/14 
2. The size of the dead zone is not large enough, which leads to the jog not settling at rest. 

Again, thanks for your effort!  If you need any help, I'll be happy to troubleshoot/debug.

winder commented on ugs-platform/ugs-platform-plugin-joystick/src/main/java/com/willwinder/ugs/nbp/joystick/model/JoystickControl.java:
```diff
@@ -0,0 +1,84 @@
+ *
+ * @author Joacim Breiler
+ */
+public enum JoystickControl {
```
This should support a lot of controllers! One thing I wonder is how it could support controllers with a digital D-Pad?

breiler replied:
Right now the support is limited. The buttons can't be mapped to a continuous jogging for instance. So if you want to jog it will only allow single stepping.

winder commented on ugs-platform/ugs-platform-plugin-joystick/src/main/java/com/willwinder/ugs/nbp/joystick/model/JoystickControl.java:
```diff
@@ -0,0 +1,84 @@
+    X("platform.plugin.joystick.x", false),
+    Y("platform.plugin.joystick.y", false),
+    BACK("platform.plugin.joystick.back", false),
+    GUIDE("platform.plugin.joystick.guide", false),
```
I've never heard of the guide button, is it the same as select? After some searching, it looks like this is an xbox thing?

It's probably safe to borrow the terminology used by retroarch. I think the changes are:
GUIDE -> SELECT
LEFT_BUMPER -> L
LEFT_TRIGGER -> L2
LEFT_STICK -> L3
RIGHT_BUMPER -> R
RIGHT_TRIGGER -> R2
RIGHT_STICK -> L3

https://retropie.org.uk/docs/RetroArch-Configuration/#example-default-per-system-retroarchcfg
https://github.com/libretro/RetroArch/blob/a96ae6e987671da71be4aa840e80eb341d4d62f3/retroarch.cfg#L491
```
# input_player1_a_btn =
# input_player1_b_btn =
# input_player1_y_btn =
# input_player1_x_btn =
# input_player1_start_btn =
# input_player1_select_btn =
# input_player1_l_btn =
# input_player1_r_btn =
# input_player1_left_btn =
# input_player1_right_btn =
# input_player1_up_btn =
# input_player1_down_btn =
# input_player1_l2_btn =
# input_player1_r2_btn =
# input_player1_l3_btn =
# input_player1_r3_btn =
```

breiler replied:
Thanks, I'll update these!

winder commented on ugs-platform/ugs-platform-plugin-joystick/src/main/java/com/willwinder/ugs/nbp/joystick/service/JoystickServiceImpl.java:
```diff
@@ -36,29 +44,88 @@ This file is part of Universal Gcode Sender (UGS).
+        joystickActionDispatcher = new ActionDispatcher(actionManager, continuousJogWorker);
+        addListener(joystickActionDispatcher);
+
+        if (Settings.getVersion() == 0) {
```
Could you add a comment here that says something like... "Check if the settings have not been initialized".

Probably not the most common edge case, but this does get a little tricky if multiple controllers are used or the controller is upgraded. It's probably fine to ignore this, since the user would be the only one who knows that the device has changed.

winder commented on ugs-platform/ugs-platform-plugin-joystick/src/main/java/com/willwinder/ugs/nbp/joystick/service/JoystickServiceImpl.java:
```diff
@@ -133,14 +209,24 @@ private void readData() {
-            // Round values, got issues with the controller having 0.01 as zero-value
-            float value = Math.round(currentController.getAxisState(controllerAxis) * 10) / 10f;
-            JoystickAxis axis = Utils.getJoystickAxisFromControllerAxis(controllerAxis);
+            // We might have rounding errors from the controller, ignore the low value range
```
Have you noticed that the analog inputs can change even if the joystick isn't moving? Would it help to keep a moving average?

winder replied:
That would probably be part of the `currentController` for analog fields.

breiler replied:
I have the threshold at 3 percent which was required for my controller. Another user reported that this wasn't enough for their controller. This should probably be configurable as well or if we could come up with a clever algorithm for this as you suggested.

winder commented on ugs-platform/ugs-platform-plugin-joystick/src/main/resources/resources/MessagesBundle.properties:
```diff
@@ -0,0 +1,26 @@
+platform.plugin.joystick.activate = Activate joystick
```
I'm still confused about how these work. Is the per-plugin message bundle overridden by the global message bundle?

breiler replied:
No you are right. I used the MessagesBundle as the module bundle before, but this doesn't work anymore with the latest version of Netbeans. I removed this.

winder commented on ugs-platform/ugs-platform-ugslib/src/main/java/com/willwinder/ugs/nbp/lib/services/ActionRegistrationService.java:
```diff
@@ -144,4 +155,87 @@ public void overrideActionName(String category, String key, String name) {
             Exceptions.printStackTrace(ex);
         }
     }
+
```
Nice! The NetBeans people might like to hear about this: https://netbeans.apache.org/wiki/DevFaqActionsAddAtRuntime.asciidoc

breiler commented on ugs-platform/ugs-platform-plugin-joystick/src/main/java/com/willwinder/ugs/nbp/joystick/Settings.java:
```diff
@@ -30,12 +31,31 @@ This file is part of Universal Gcode Sender (UGS).
     public static final String SETTINGS_ACTIVE = "active";
+    public static final String SETTINGS_VERSION = "version";
 
     private static Preferences preferences = NbPreferences.forModule(JoystickService.class);
```
Yes they are persisted. The downside is that the "Apply" button has little use as the settings are made instantly. But I figure that this can be added later when we have tried these features.

breiler commented on ugs-platform/ugs-platform-plugin-joystick/src/main/java/com/willwinder/ugs/nbp/joystick/Utils.java:
```diff
@@ -20,54 +20,66 @@ This file is part of Universal Gcode Sender (UGS).
      * @return a joystick button if it could be translated, or throws an illegalargumentexception
      */
-    public static JoystickButton getJoystickButtonFromControllerButton(ControllerButton controllerButton) {
+    public static JoystickControl getJoystickButtonFromControllerButton(ControllerButton controllerButton) {
```
I just wanted to separate our code from the Jamepad. This way we could can change the library without changing any of our implementation.

breiler commented:
@jiffyjeff thanks for trying this out! I made some updates:

> 1. The Y and Z axis directions were reversed for the analog joysticks. 

Added the possibility to reverse the axises.

> 2. The size of the dead zone is not large enough, which leads to the jog not settling at rest.

Added a setting for the zero threshold.
